File: Daily Challenge/February 2025/Product of the Last K Numbers.cpp
class ProductOfNumbers {
private:
  long long product = 1;
  vector<long long> prefix = {1};
public:
  ProductOfNumbers() {
    
  }
  
  void add(int num) {
    if (num == 0) {
      prefix = {1};
      product = 1;
    } else {
      product *= num;
      prefix.push_back(product);
    }
  }
  
  int getProduct(int k) {
    if (k >= prefix.size()) {
      return 0;
    }
    
    return prefix.back() / prefix[prefix.size() - k - 1];
  }
};
  
  /**
   * Your ProductOfNumbers object will be instantiated and called as such:
   * ProductOfNumbers* obj = new ProductOfNumbers();
   * obj->add(num);
   * int param_2 = obj->getProduct(k);
   */

File: Daily Challenge/February 2025/Product of the Last K Numbers.cs
public class ProductOfNumbers {

  private List<int> products;

  public ProductOfNumbers() {
    products = new List<int>();
    products.Add(1);
  }
  
  public void Add(int num) {
    if (num == 0) {
      products = new List<int>();
      products.Add(1);
    } else {
      products.Add(products.Last() * num);
    }
  }
  
  public int GetProduct(int k) {
    if (k >= products.Count) {
      return 0;
    }
    return products.Last() / products[products.Count - k - 1];
  }
}

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers obj = new ProductOfNumbers();
 * obj.Add(num);
 * int param_2 = obj.GetProduct(k);
 */